dbms/src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -303,81 +314,101 @@ InterpreterSelectQuery::InterpreterSelectQuery(
+                            table_expression,
+                            getSubqueryContext(context),
+                            options.subquery(),
+                            required_columns);
+            }
         }
-    }
-
-    if (interpreter_subquery)
-    {
-        /// If there is an aggregation in the outer query, WITH TOTALS is ignored in the subquery.
-        if (query_analyzer->hasAggregation())
-            interpreter_subquery->ignoreWithTotals();
-    }
 
-    required_columns = syntax_analyzer_result->requiredSourceColumns();
+        if (interpreter_subquery)
+        {
+            /// If there is an aggregation in the outer query, WITH TOTALS is ignored in the subquery.
+            if (query_analyzer->hasAggregation())
+                interpreter_subquery->ignoreWithTotals();
+        }
 
-    if (storage)
-    {
-        source_header = storage->getSampleBlockForColumns(required_columns);
+        required_columns = syntax_analyzer_result->requiredSourceColumns();
 
-        /// Fix source_header for filter actions.
-        if (context.hasUserProperty(storage->getDatabaseName(), storage->getTableName(), "filter"))
+        if (storage)
         {
-            filter_info = std::make_shared<FilterInfo>();
-            filter_info->column_name = generateFilterActions(filter_info->actions, storage, context, required_columns);
-            source_header = storage->getSampleBlockForColumns(filter_info->actions->getRequiredColumns());
+            source_header = storage->getSampleBlockForColumns(required_columns);
+
+            /// Fix source_header for filter actions.
+            if (context.hasUserProperty(storage->getDatabaseName(), storage->getTableName(), "filter"))
+            {
+                filter_info = std::make_shared<FilterInfo>();
+                filter_info->column_name = generateFilterActions(filter_info->actions, storage, context, required_columns);
+                source_header = storage->getSampleBlockForColumns(filter_info->actions->getRequiredColumns());
+            }
         }
-    }
 
-    /// Calculate structure of the result.
-    result_header = getSampleBlockImpl();
-    for (auto & col : result_header)
+        /// Calculate structure of the result.
+        result_header = getSampleBlockImpl();
+    };
+
+    analyze();
+
+    bool need_analyze_again = false;
+    if (analysis_result.prewhere_constant_filter_description.always_false || analysis_result.prewhere_constant_filter_description.always_true)
     {
-        if (!col.column)
-            col.column = col.type->createColumn();
-        else if (isColumnConst(*col.column) && !col.column->empty())
-            col.column = col.column->cloneEmpty();
+        auto constant = std::make_shared<ASTLiteral>(0u);
+        if (analysis_result.prewhere_constant_filter_description.always_true)
+            constant->value = 1u;
+        query.refPrewhere() = constant;
```

Comment:
```ref``` functions  are candidates to remove. They are still alive cause it's hard to remove them in some not-const visitors. There's no guaranty that setting the value you remove old AST node from query children (it's iplementation specific). Use ```setExpression(Expression::PREWHERE, constant)``` instead.